File: src/Day01.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Day01 {
    static final String FILENAME = "./src/inputs/Day01.txt";

    static List<Integer> expenseReport = Utils.readLines(FILENAME).stream().map(Integer::parseInt).collect(Collectors.toList());

    static int part01() {
        var expenseList = new ArrayList<Integer>();
        for (var expense : expenseReport) {
            var diff = Math.abs(2020 - expense);
            if (expenseList.contains(diff))
                return expense * diff;
            else
                expenseList.add(expense);
        }
        return -1;
    }

    static int part02() {
        for (var i = 0; i < expenseReport.size(); i++) {
            for (var j = i + 1; j < expenseReport.size(); j++) {
                var expenseOnePlusTwo = expenseReport.get(i) + expenseReport.get(j);
                var diff = 2020 - expenseOnePlusTwo;
                if (!(expenseOnePlusTwo < 2020) && (diff < 0))
                    continue;
                if (expenseReport.contains(diff))
                    return diff * expenseReport.get(i) * expenseReport.get(j);
            }
        }
        return -1;
    }

    public static void main() {
        System.out.println("Advent of Code 2020, Day 01");
        System.out.println(part01());
        System.out.println(part02());
    }
}

File: day01/day01.go
package day01

import (
	"advent-of-code-2020/utils"
	"math"
	"slices"
)

const filename = "./inputs/day01.txt"

var expenseReport = utils.MapToInt(utils.ReadLines(filename))

func part01() int {
	var expenseList []int
	for _, expense := range expenseReport {
		diff := int(math.Abs(2020 - float64(expense)))
		if slices.Contains(expenseList, diff) {
			return expense * diff
		}
		expenseList = append(expenseList, expense)
	}
	return -1
}

func part02() int {
	for i := 0; i < len(expenseReport)-1; i++ {
		for j := i + 1; j < len(expenseReport); j++ {
			expenseOnePlusTwo := expenseReport[i] + expenseReport[j]
			diff := 2020 - expenseOnePlusTwo
			if !(expenseOnePlusTwo < 2020) && (diff < 0) {
				continue
			}
			if slices.Contains(expenseReport, diff) {
				return diff * expenseReport[i] * expenseReport[j]
			}
		}
	}
	return -1
}

func Main() (int, func() int, func() int) {
	return 1, part01, part02
}
